Adding a --update command for pulling new target defintions from online

bridadan opened the pull request:
This is a very basic implementation of #119.

FYI @theotherjimmy @jupe 

**TODO**

- [x] Indicate success of update with error code
- [x] Throw exception on failed update (edit: using a return code)
- [x] Add doc string for public api
- [x] Document structure of JSON for plat_db
- [x] Add url argument for the update cli command
~~- [ ] Change the `--update` argument to a full subcommand (ex `mbedls update`)~~
        - Edit: holding off on this for a later PR
- [x] Add more log output during the update

theotherjimmy commented:
:balloon: :tada: :birthday:

jupe commented on mbed_lstools/main.py:
```diff
@@ -130,6 +130,9 @@ def mock_platform(mbeds, args):
             logger.error("Could not parse mock from token: '%s'", token)
 
+def update_from_web(mbeds, args):
+    mbeds.plat_db.update_from_web()
```
would be good to indicate if update success or not as a retcode

theotherjimmy replied:
Agreed

jupe commented on mbed_lstools/platform_database.py:
```diff
@@ -478,3 +479,16 @@ def remove(self, id, permanent=False, device_type='daplink', verbose_data=False)
 
                 return _modify_data_format(removed, verbose_data)
+
+    def update_from_web(self, url='https://os.mbed.com/api/v3/platforms/?format=json'):
```
is this public API ? this could throw some specific exception when updating fails for some reason

theotherjimmy replied:
@bridadan This is a public API, and it should have a docstring, yeah?

bridadan replied:
Yes to both! I'll make a little todo in the original post

jupe commented on test/platform_database.py:
```diff
@@ -149,6 +149,26 @@ def test_simplify_verbose_data(self):
+            _result = MagicMock()
+            _result.configure_mock(**{
+                'json.return_value': [
+                    {
```
maybe could be good to document this json format

theotherjimmy replied:
Good point. FYI, this is a test.

jupe commented on mbed_lstools/main.py:
```diff
@@ -204,6 +207,9 @@ def parse_cli(to_parse):
         '-m', '--mock', metavar='ID:NAME',
         help='substitute or create a target ID to platform name mapping used'
         'when invoking mbedls in the current directory')
+    commands.add_argument(
```
maybe here could be also optional update url argument that is used to fetch updates - with default value which are visible in help..

theotherjimmy replied:
Definitely. Very good spot!

bridadan commented:
Great comments! I'll address those in the next round of updates.

One thing I talked with @theotherjimmy offline is making our "subcommands" (mock, list platforms, update) actual command line "subcommands" instead of arguments. I think this PR could probably introduce that concept first and we could address the other commands in a separate PR.

This would give us the ability to fix #251.

bridadan commented:
After looking into it a bit, argparse doesn't support "optional subparsers". Meaning we can't have `mbedls` list mbeds while still having the `mbedls update` command. We'd need to move to something like [docopt](https://github.com/docopt/docopt) or another CLI library, which should be done in another PR ahead of this. What do you think @theotherjimmy ?

theotherjimmy commented:
@bridadan Agreed. Add switches for now. We'll prototype mbed-ls 2.0's sub-command style in a future PR.

bridadan commented:
Ok, I've added quite a bit of output. New platforms are now printed in a table. Updated platform names are printed like `ARM_MPS2_M0 -> ARM_IOTSS_BEID`. An example is shown below:


```
$ mbedls --update
WARNING:mbedls.platform_database:Error loading database C:\Users\bridan01\AppData\Local\mbedls\mbedls\platforms.json: [Errno 22] Invalid argument; Recreating
Fetching platform data from https://os.mbed.com/api/v3/platforms/?format=json...
+-----------+-------------------------------+
| target_id | platform_name                 |
+-----------+-------------------------------+
| 0320      | MTS_MDOT_F411RE               |
| 5001      | ARM_MPS2_M0 -> ARM_IOTSS_BEID |
| 5004      | ARM_CM3DS_MPS2                |
| 7401      | LPC11U35_501                  |
| 7405      | LPC11U35_501                  |
| 9013      | ARCH_LINK                     |
| C000      | LPC812                        |
+-----------+-------------------------------+
The platform database was updated.
```

When no new platforms were found, this is the output:

```
$ mbedls --update
Fetching platform data from https://os.mbed.com/api/v3/platforms/?format=json...
Platform database up-to-date.
```

To accomplish this, I've changed the return value of the `update_from_web` function. The output is detailed in the function's docstring. For errors that occur, it will now raised those as exceptions: `RemotePlatformDataException` for web related issues and `DatabaseUpdateException` for filesystem related issues.

bridadan commented:
Aaaaand da tests are in 👌

jupe commented on mbed_lstools/platform_database.py:
```diff
@@ -480,15 +489,101 @@ def remove(self, id, permanent=False, device_type='daplink', verbose_data=False)
             self.add(id, platform_name)
 
-        self._update_db()
+        if not self._update_db():
```
why update_db does not raise exception ?

theotherjimmy replied:
I'm not seeing this part of the diff.

bridadan replied:
It never has

jupe commented on mbed_lstools/platform_database.py:
```diff
@@ -480,15 +489,101 @@ def remove(self, id, permanent=False, device_type='daplink', verbose_data=False)
+
+        Returns dictionary of all updates that occurred in the following format:
+            {
+                "new": {
```
why this function could not overwrite whole local db? isn’t this try to do same thing anyway..?

theotherjimmy replied:
To avoid blowing away the jlink info.

jupe commented on mbed_lstools/platform_database.py:
```diff
@@ -480,15 +489,101 @@ def remove(self, id, permanent=False, device_type='daplink', verbose_data=False)
+                {
+                    "productcode": "1234",
+                    "logicalboard": {
+                        "name": "boardname"
```
is boardname == platform_name ?

theotherjimmy replied:
Yes

jupe commented on mbed_lstools/platform_database.py:
```diff
@@ -480,15 +489,101 @@ def remove(self, id, permanent=False, device_type='daplink', verbose_data=False)
         r = requests.get(url)
-        platform_data = r.json()
+        try:
+            platform_data = r.json()
```
would be good if received json blob is schema validated before using json to avoid unexpected behaviour..

theotherjimmy replied:
Next PR?

bridadan replied:
I do a bit of catching of schema errors on line 541. It prints to debug if the schema changes. I don't think we'll be doing anything stronger than that, except maybe raising a warning?

jupe commented:
ups, seems that I was a bit too late for new comments 😁

theotherjimmy commented:
@jupe It's not released just yet. Comments are still welcome.
